epochStart/metachain/systemSCs.go

```diff
@@ -327,9 +333,134 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
 		}
 	}
 
+	if s.flagStakingV4Enabled.IsSet() {
+		allNodesKeys := s.getAllNodeKeys(validatorInfos)
+
+		err := s.stakingDataProvider.PrepareStakingData(allNodesKeys)
+		if err != nil {
+			return err
+		}
+
+		err = s.selectNodesFromAuctionList(validatorInfos, randomness)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *systemSCProcessor) selectNodesFromAuctionList(validatorInfos map[uint32][]*state.ValidatorInfo, randomness []byte) error {
+	auctionList, noOfValidators := getAuctionListAndNoOfValidators(validatorInfos)
+	err := s.sortAuctionList(auctionList, randomness)
+	if err != nil {
+		return err
+	}
+
+	auctionListSize := uint32(len(auctionList))
+	noOfAvailableNodeSlots := core.MinUint32(auctionListSize, s.maxNodes-noOfValidators)
+	s.displayAuctionList(auctionList, noOfAvailableNodeSlots)
+
+	for i := uint32(0); i < noOfAvailableNodeSlots; i++ {
+		auctionList[i].List = string(common.NewList)
+	}
+
+	return nil
+}
+
+func getAuctionListAndNoOfValidators(validatorInfos map[uint32][]*state.ValidatorInfo) ([]*state.ValidatorInfo, uint32) {
+	auctionList := make([]*state.ValidatorInfo, 0)
+	noOfValidators := uint32(0)
+
+	for _, validatorsInShard := range validatorInfos {
+		for _, validator := range validatorsInShard {
+			if validator.List == string(common.AuctionList) {
+				auctionList = append(auctionList, validator)
+			} else if isValidator(validator) {
+				noOfValidators++
+			}
+		}
+	}
+
+	return auctionList, noOfValidators
+}
+
+func (s *systemSCProcessor) sortAuctionList(auctionList []*state.ValidatorInfo, randomness []byte) error {
+	errors := make([]error, 0)
+
+	sort.SliceStable(auctionList, func(i, j int) bool {
+		pubKey1 := auctionList[i].PublicKey
+		pubKey2 := auctionList[j].PublicKey
+
+		nodeTopUpPubKey1, err := s.stakingDataProvider.GetNodeStakedTopUp(pubKey1)
```

Comment:
this is sub-optimal as we can call multiple times the GetNodeStakedTopUp on the same public key.
Why not create a map where the key is the public key and the value is the top up value on L390? (outside of this function)
This function will only read from that constructed map.

Reply:
Right; very good suggestion.
I've done in this because I've seen that the data is taken from an internal staking cache (`ownerInfo, ok := sdp.cache[owner]`), but apparently it only does that after calling a smart contract (`"getOwner"`).

I've created a new func: `getNodeTopUpValidatorMap` which returns the suggested map: `map[string]*big.Int`